Python/Product/Analysis/LanguageServer/OpenFiles.cs

```diff
@@ -0,0 +1,132 @@
+            if (changes == null) {
+                return;
+            }
+
+            var uri = @params.textDocument.uri;
+            var doc = _projectFiles.GetEntry(uri) as IDocument;
+            if (doc == null) {
+                return;
+            }
+
+            _delayedAction?.Cancel();
+
+            _documentChangeProcessingComplete.Reset();
+            _documentChangeReentrancyCount++;
+
+            try {
+                var part = _projectFiles.GetPart(uri);
+                _log.TraceMessage($"Received changes for {uri}");
+
+                var docVersion = Math.Max(doc.GetDocumentVersion(part), 0);
+                var fromVersion = Math.Max(@params.textDocument.version - 1 ?? docVersion, 0);
+
+                if (fromVersion > docVersion && @params.contentChanges?.Any(c => c.range == null) != true) {
+                    // Expected from version hasn't been seen yet, and there are no resets in this
+                    // change, so enqueue it for later.
+                    _log.TraceMessage($"Deferring changes for {uri} until version {fromVersion} is seen");
+                    lock (PendingChanges) {
+                        PendingChanges.Add(@params);
+                    }
+                    return;
+                }
+
+                var toVersion = @params.textDocument.version ?? (fromVersion + changes.Length);
+
+                doc.UpdateDocument(part, new DocumentChangeSet(
+                    fromVersion,
+                    toVersion,
+                    changes.Select(c => new DocumentChange {
+                        ReplacedSpan = c.range.GetValueOrDefault(),
+                        WholeBuffer = !c.range.HasValue,
+                        InsertedText = c.text
+                    })
+                ));
+
+                DidChangeTextDocumentParams? next = null;
+                lock (PendingChanges) {
+                    var notExpired = PendingChanges
+                        .Where(p => p.textDocument.version.GetValueOrDefault() >= toVersion)
+                        .OrderBy(p => p.textDocument.version.GetValueOrDefault())
+                        .ToArray();
+
+                    PendingChanges.Clear();
+                    if (notExpired.Any()) {
+                        next = notExpired.First();
+                        PendingChanges.AddRange(notExpired.Skip(1));
+                    }
+                }
+                if (next.HasValue) {
+                    DidChangeTextDocument(next.Value, analysisDelay, enqueueAction);
+                }
+            } finally {
+                _documentChangeReentrancyCount--;
+                if (_documentChangeReentrancyCount == 0) {
+
+                    _log.TraceMessage($"Applied changes to {uri}");
+                    if (analysisDelay > 0) {
+                        _delayedAction = new DelayedAction(() => enqueueAction(doc), analysisDelay);
+                    } else {
+                        enqueueAction(doc);
```

Comment:
The problem I've been getting is that as soon unit is enqueued, current module analysis becomes empty as variables are cleared. Specifically https://github.com/Microsoft/PTVS/blob/master/Python/Product/Analysis/ProjectEntry.cs@234

Reply:
With the most recent change after dot typed code find the proper element since it is provided with the `_expr` hint like VS does.

Reply:
> as soon unit is enqueued, current module analysis becomes empty as variables are cleared

These are shallow-copied when we create the `ModuleAnalysis` at the end of that function, so clearing them shouldn't be an issue. What we probably need to do is not publish the new analysis if it was only enqueued (unless we don't have one yet, in which case querying the in-progress one is better than nothing).

Reply:
Shallow copy is not sufficient. The problem is that variables lose types. So even if we clone dictionaries (I added that), variables still lose types. We could declare something like `ICloneable` and if dictionary value implements it, invoke the clone. HTML works this way.

Also, scope is currently cloned when new analysis is created but it gets cleared when unit is queued for the analysis. We could create new `ModuleInfo` rather than reusing existing one.